### ISSUE TITLE
How to perform an async validation? it's missing from docs

### ISSUE THREAD

**Issue by leosuncin** [smikhalevski/doubter#80]
The README did mention that it's possible to create async validations but it doesn't include an example, neither the tests include one.

I tried to use `check` method

```ts
import * as d from 'doubter';

async function isOdd(value: number): Promise<boolean> {
  return value % 2 === 0;
}

const oddNumber = d.number()
  .coerce()
  .check(async (value) => {
    if (await isOdd(value)) {
      return null;
    }
    return {
      code: 'odd_number',
      message: 'Number must be odd',
    } satisfies d.Issue;
  });

await oddNumber.parseAsync(4);
```

<details>
<summary>⚠️ Typescript error</summary>
<pre>
error TS2345: Argument of type '(value: number) => Promise<{ code: string; message: string; } | null>' is not assignable to parameter of type 'CheckCallback<number, undefined>'.
Type 'Promise<{ code: string; message: string; } | null>' is not assignable to type 'void | Issue | Issue[] | null | undefined'.
</pre>
</details>

<details>
<summary>💥 Runtime error</summary>
<pre>
ValidationError: [
  {}
]
    at new ValidationError (/Users/developer/Projects/express-mssql-example/node_modules/.pnpm/doubter@2.1.0/node_modules/doubter/lib/index.js:370:28) {
  issues: [
    Promise {
      undefined,
      [Symbol(async_id_symbol)]: 909,
      [Symbol(trigger_async_id_symbol)]: 6
    }
  ]
}
</pre>
</details>

Or using `refine` method

```ts
import * as d from 'doubter';

async function isOdd(value: number): Promise<boolean> {
  return value % 2 === 0;
}

const oddNumber = d.number()
  .coerce()
  .refine(isOdd);
  
await oddNumber.parseAsync(3)
```

<details>
<summary>⚠️ Typescript error</summary>
<pre>
error TS2769: No overload matches this call.
Overload 1 of 2, '(cb: (output: number, options: Readonly<ApplyOptions>) => output is number, options?: RefineOptions | Message | undefined): Shape<number, number>', gave the following error.
  Argument of type '(value: number) => Promise<boolean>' is not assignable to parameter of type '(output: number, options: Readonly<ApplyOptions>) => output is number'.
    Signature '(value: number): Promise<boolean>' must be a type predicate.
Overload 2 of 2, '(cb: (output: number, options: Readonly<ApplyOptions>) => boolean, options?: RefineOptions | Message | undefined): NumberShape', gave the following error.
  Argument of type '(value: number) => Promise<boolean>' is not assignable to parameter of type '(output: number, options: Readonly<ApplyOptions>) => boolean'.
    Type 'Promise<boolean>' is not assignable to type 'boolean'.
</pre>
</details>

<details>
<summary>🤷 no error but it always pass</summary>
<pre>
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 1254,
  [Symbol(trigger_async_id_symbol)]: 1253
}
3
</pre>
</details>

**Doubter version:** 2.1.0
**Node.js version:** 20.3.1

PS: my actual use case is that I want to check whether a value exist in the DB

**Comment by smikhalevski** [comment]
Checks are synchronous. You can use `transformAsync` and throw a `ValidationError`:

```ts
import * as d from 'doubter';

async function isOdd(value: number): Promise<boolean> {
  return value % 2 === 0;
}

const oddNumber = d.number()
  .coerce()
  .transformAsync(async (value) => {
    if (await isOdd(value)) {
      return value;
    }

    throw new d.ValidationError([
      {
        code: 'odd_number',
        message: 'Number must be odd',
      }
    ]);
  });

await oddNumber.parseAsync(4);
```

**Comment by leosuncin** [comment]
In that case, a new function called `checkAsync` would be a better fit since it communicates the intention of doing an asynchronous verification, the same goes for another new function `refineAsync` and make `refine` to only accept sync functions

**Comment by smikhalevski** [comment]
I really like this idea! I would try to add support for async [operations](https://github.com/smikhalevski/doubter#operations) in the next minor release of v3.

**Comment by smikhalevski** [comment]
Async operations, checks, refinements, and alterations are now available in [v4.0.0](https://github.com/smikhalevski/doubter/releases/tag/v4.0.0).
